pytensor/tensor/blockwise.py

```diff
@@ -308,91 +407,62 @@ def L_op(self, inputs, outs, ograds):
-            inner_outputs_storage = [[None] for _ in range(n_outs)]
-
-            def core_func(
-                *inner_inputs,
-                core_node=core_node,
-                inner_outputs_storage=inner_outputs_storage,
-            ):
-                self.core_op.perform(
-                    core_node,
-                    [np.asarray(inp) for inp in inner_inputs],
-                    inner_outputs_storage,
-                )
+            if len(node.outputs) == 1:
 
-                if n_outs == 1:
-                    return inner_outputs_storage[0][0]
-                else:
-                    return tuple(r[0] for r in inner_outputs_storage)
+                def gufunc(*inputs):
+                    _check_runtime_broadcast(inputs, batch_bcast_patterns, batch_ndim)
+                    return (core_func(*inputs),)
+            else:
 
-            gufunc = np.vectorize(core_func, signature=self.signature)
+                def gufunc(*inputs):
+                    _check_runtime_broadcast(inputs, batch_bcast_patterns, batch_ndim)
+                    return core_func(*inputs)
+        else:
+            core_node = self._create_dummy_core_node(node.inputs)  # type: ignore
+            gufunc = _vectorize_node_perform(
+                core_node,
+                batch_bcast_patterns=batch_bcast_patterns,
+                batch_ndim=self.batch_ndim(node),
+                impl=impl,
+            )
 
-        node.tag.gufunc = gufunc
+        return gufunc
 
     def _check_runtime_broadcast(self, node, inputs):
         batch_ndim = self.batch_ndim(node)
+        batch_bcast = [pt_inp.type.broadcastable[:batch_ndim] for pt_inp in node.inputs]
+        _check_runtime_broadcast(inputs, batch_bcast, batch_ndim)
 
-        # strict=False because we are in a hot loop
-        for dims_and_bcast in zip(
-            *[
-                zip(
-                    input.shape[:batch_ndim],
-                    sinput.type.broadcastable[:batch_ndim],
-                    strict=False,
-                )
-                for input, sinput in zip(inputs, node.inputs, strict=False)
-            ],
-            strict=False,
-        ):
-            if any(d != 1 for d, _ in dims_and_bcast) and (1, False) in dims_and_bcast:
-                raise ValueError(
-                    "Runtime broadcasting not allowed. "
-                    "At least one input has a distinct batch dimension length of 1, but was not marked as broadcastable.\n"
-                    "If broadcasting was intended, use `specify_broadcastable` on the relevant input."
-                )
+    def prepare_node(self, node, storage_map, compute_map, impl=None):
+        node.tag.gufunc = self._create_node_gufunc(node, impl=impl)
 
     def perform(self, node, inputs, output_storage):
-        gufunc = getattr(node.tag, "gufunc", None)
-
-        if gufunc is None:
-            # Cache it once per node
-            self._create_node_gufunc(node)
+        try:
```

Comment:
```py
gu_func = node.tag.gufunc if hasattr(node.tag, 'gufunc') else self.create_node_gufunc(node, impl=None)
```
?

Reply:
Slower!

Reply:
```python
class A:
    pass

def create_f(a):
    return 0

def f(a):
    f = a.f if hasattr(a, "f") else create_f(a)
    
def g(a):
    try:
        f = a.f
    except AttributeError:
        f = a.f = create_f(a)

a = A()  # 129 ns ± 33.2 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
%timeit f(a)

a = A()
%timeit g(a)  # 46.1 ns ± 2.7 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
that's frankly shocking. do you know why try/expect is faster?

Reply:
attribute accesses are slow, and with the hasattr you have twice as many in every call. Also in this case we know the create_f is only called once, so we expect it to work most of the times